File: challenge2_palindrome_checker/python/solution_python/dequeue.py
import re

class Dequeue:
    def __init__(self):
        self._items = []

    def set_items(self, string):
        self._items = list(string)

    def add_front(self, item):
        self._items.insert(0, item)

    def add_rear(self, item):
        self._items.append(item)

    def remove_front(self):
        if self._items:
           return self._items.pop(0)

    def remove_rear(self):
        if self._items:
            return self._items.pop()

    def peek_front(self):
        if self._items:
            return self._items[0]

    def peek_rear(self):
        if self._items:
            return self._items[-1]

    def size(self):
        return len(self._items)
        
    def is_empty(self):
        return len(self._items) == 0
 


    

File: challenge2_palindrome_checker/php/challenge2_extension_php/Dequeue.php
<?php

class Dequeue
{
    protected $items;

    public function __construct()
    {
        $this->items = [];
    }

    public function setItems($string)
    {
        $items = str_split($string);

        foreach ($items as $item) {
            if (ord($item) >= 97 and ord($item) <= 122) {
                $this->addRear($item);
            } elseif (ord($item) >= 65 and ord($item) <= 90) {
                // difference between upper case and lowercase ascii valies is +32
                $this->addRear(chr(ord($item) + 32));
            }
        }
    }

    public function addFront($item)
    {
        array_unshift($this->items, $item);
    }

    public function addRear($item)
    {
        array_push($this->items, $item);
    }

    public function removeFront()
    {
        return array_shift($this->items);
    }

    public function removeRear()
    {
        return array_pop($this->items);
    }

    public function peekFront()
    {
        return $this->items[0];
    }

    public function peekRear()
    {
        return end($this->items);
    }

    public function size()
    {
        return count($this->items);
    }

    public function isEmpty()
    {
        return $this->size() == 0;
    }
}